Add audiocore.MixerVoice

sommersoft opened the pull request:
Fixes #1306. Also possible that this address #1260 & #1305.

Tested with:
- WaveFile: 8bit (unsigned) & 16bit (signed)
- RawSample: 8 & 16 bit (signed & unsigned)

ItsyBitsy M0 Express (-360 bytes) & Metro M0 Express (-4 bytes) are failing Pinyin builds on my Travis. Open to suggestions on how to handle those.

--------------------------------
Notes for Release:
---------------------
- Creates new `audiomixer` module. 
- `audioio.Mixer` is now `audiomixer.Mixer`, but is backwards compatible where `audiomixer` is available (for now).
- `audiomixer` is not available on M0 (SAMD21) based boards.
- Adds `MixerVoice` instance objects for `Mixer` voices, and provides independent level adjustment for each voice (0.0-1.0; amplification is not available).

dhalbert commented on shared-bindings/audiocore/MixerVoice.c:
```diff
@@ -0,0 +1,221 @@
+        { MP_QSTR_voice, MP_ARG_INT, {.u_int = 0} },
+    };
+    audioio_mixervoice_obj_t *self = MP_OBJ_TO_PTR(pos_args[0]);
+    //raise_error_if_deinited(common_hal_audioio_mixer_deinited(self));
```
Did you mean for this to be commented out?

sommersoft replied:
Doh! I meant to go back and fix those, as they were causing build errors (implicit declaration).

sommersoft replied:
And in looking closer, `deinit` as a whole needs work... 🤦‍♂

tannewt replied:
I'd just remove the deinit stuff since the only shared resource this module uses is memory.

dhalbert commented on shared-bindings/audiocore/MixerVoice.c:
```diff
@@ -0,0 +1,221 @@
+//|
+STATIC mp_obj_t audioio_mixervoice_obj_get_level(mp_obj_t self_in) {
+
+	#if !MICROPY_PY_BUILTINS_FLOAT
```
Don't need this - we'll never do a port without floats.

dhalbert commented on shared-bindings/audiocore/MixerVoice.c:
```diff
@@ -0,0 +1,221 @@
+    mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
+    mp_arg_parse_all(n_args - 1, pos_args + 1, kw_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
+
+	#if MICROPY_PY_BUILTINS_FLOAT
```
same

dhalbert commented on shared-module/audiocore/Mixer.c:
```diff
@@ -152,31 +116,31 @@ uint32_t add8signed(uint32_t a, uint32_t b) {
 
 uint32_t add8unsigned(uint32_t a, uint32_t b) {
-    #if (defined (__ARM_FEATURE_DSP) && (__ARM_FEATURE_DSP == 1))
+    /*#if (defined (__ARM_FEATURE_DSP) && (__ARM_FEATURE_DSP == 1))
```
Why did you give up on the saturating arithmetic builtins? They are tiny and will help save space.

sommersoft replied:
In short: I don't currently understand it enough to accomplish it. These were originally done by DeanM, but only the 16 bit signed funcs were complete. The rest were mostly stubs/initial.

dhalbert replied:
OK, looking at this in more detail, the saturated builtins are only available on M4, not M0. There's a lot of stuff online about how to implement unsigned and signed saturated arithmetic. This is good (original web page is blank at the moment): https://webcache.googleusercontent.com/search?q=cache:bSrCZnMkEfkJ:https://locklessinc.com/articles/sat_arithmetic/+&cd=3&hl=en&ct=clnk&gl=us

There are a few example libraries in github but mostly in C++, unfortunately. The code here could maybe be adapted for 8 and 16 bits: https://github.com/esote/bsa/blob/master/bsa.c
https://stackoverflow.com/questions/121240/how-to-do-unsigned-saturating-addition-in-c
https://stackoverflow.com/questions/17580118/signed-saturated-add-of-64-bit-ints
etc.

ladyada replied:
i srsly doubt this would fit in m0 anyways, lets have it for m4 only!

dhalbert replied:
if it's only for m4, you can just use the builtins and not bother to figure out other implementations. Just throw a compile `#error` for now or something like that.

sommersoft replied:
The only M0 builds that include this are express boards with `audiocore` turned on (via `audioio`). 

As of now, there are 5 build failures; all 5 only fail the Pinyin translated versions. I used a Feather M0 Express during dev/testing. The en_US build currently has `4168 bytes free in flash` (in my env).

If not including `MixerVoice` on M0 is really how we want to proceed, it will require maintaining the current _Mixer only_ code for M0, and the _Mixer+MixerVoice_ for the M4 (using `#ifdefs`). Otherwise, it'll all need to be moved out of `audiocore` to gain a more granular control.

tannewt replied:
I'm ok either turning of underused modules like bitbangio or i2cslave to make this fit. It'd also be ok to move it to `audiomixer` module to give ourselves flexibility. (I want to do `audiosynth` at some point too.)

sommersoft replied:
New state of things accounting for all of this, plus the move to `audiomixer`:

- `audiomixer` is turned off completely for M0. This is a breaking change to `audioio.Mixer`.
- ARMv7M intrinsic functions are used where available (all addition functions, and 16bit signed multiplication). Also changed the pre-processor flags; the intrinsics used are in CMSIS-M, not CMSIS-DSP.
- C functions are used where intrinsic isn't available. They are also available as alternates to supported ARMv7M intrinsics, should they be desired in the future. Dan pointed me to some gcc builtins that may help slim them down (https://gcc.gnu.org/onlinedocs/gcc/Integer-Overflow-Builtins.html), which I plan on testing.
- Made better `TODO`s, I hope.

siddacious commented:
👀

tannewt commented on shared-bindings/audiocore/Mixer.c:
```diff
@@ -146,54 +163,6 @@ STATIC mp_obj_t audioio_mixer_obj___exit__(size_t n_args, const mp_obj_t *args)
-//|   .. method:: stop_voice(voice=0)
-//|
-//|     Stops playback of the sample on the given voice.
-//|
```
Do we want to leave these two functions in for backwards compatibility? They can make the equivalent MixerVoice calls.

sommersoft replied:
I can try putting them back in. With the associated space-constraints above, I'll probably approach it with use of `#if CIRCUITPY_AUDIOIO_COMPAT` and `#if CIRCUITPY_AUDIOMIXER` to not further introduce running out of space.

sommersoft replied:
This is working within the context of `audiomixer`. But, since moving to a separate module, it will still require updates to old code using `audioio`.

sommersoft replied:
> But, since moving to a separate module, it will still require updates to old code using audioio.

Scratch this. `audioio` will still work for boards that include `audiomixer`; see https://github.com/adafruit/circuitpython/pull/2080#discussion_r319523516

tannewt commented on shared-bindings/audiocore/MixerVoice.c:
```diff
@@ -0,0 +1,221 @@
+STATIC mp_obj_t audioio_mixervoice_obj___exit__(size_t n_args, const mp_obj_t *args) {
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(audioio_mixervoice___exit___obj, 4, 4, audioio_mixervoice_obj___exit__);
```
We can remove these. This lifecycle management doesn't make sense because no hardware is used directly.

tannewt commented on shared-bindings/audiocore/MixerVoice.c:
```diff
@@ -0,0 +1,221 @@
+	audioio_mixervoice_obj_t *self = m_new_obj(audioio_mixervoice_obj_t);
+    self->base.type = &audioio_mixervoice_type;
+    self->sample = NULL;
+    self->level = ((1 << 15)-1);
```
Please call out to a shared_module construct method here.

tannewt commented on shared-bindings/audiocore/MixerVoice.c:
```diff
@@ -0,0 +1,221 @@
+}
+MP_DEFINE_CONST_FUN_OBJ_KW(audioio_mixervoice_stop_obj, 1, audioio_mixervoice_obj_stop);
+
+//|   .. method:: level()
```
This should be documented as an attribute (aka property.)

tannewt commented on shared-module/audiocore/Mixer.c:
```diff
@@ -126,9 +88,11 @@ bool common_hal_audioio_mixer_get_playing(audioio_mixer_obj_t* self) {
 void audioio_mixer_reset_buffer(audioio_mixer_obj_t* self,
                                 bool single_channel,
                                 uint8_t channel) {
+#if 0
```
Remove

sommersoft commented:
Fixed most of the above, and moved Mixer to a new `audiomixer` module. After this Travis build, I'll run through the failures and disable accordingly.

> Please let us know if you'd like to hand it off.

I'm happy to push it through to the finish. But, if it's time-sensitive, no problem with passing the baton. I feel I can finish the ASM/CMSIS math functions; just takes time for me to bolster my understanding.

sommersoft commented on shared-bindings/audiomixer/Mixer.c:
```diff
@@ -36,25 +38,31 @@
-//|   :param int channel_count: The maximum number of samples to mix at once
+//|   :param int voice_count: The maximum number of voices to mix
 //|   :param int buffer_size: The total size in bytes of the buffers to mix into
+//|   :param int channel_count: The maximum number of samples to mix at once
```
Yes, `channel_count` is mono v stereo. That doc string definitely needs to be changed.

`RawSample` docs say this: "channel_count (int) – The number of channels in the buffer". But has a longer explanation in the class doc. https://circuitpython.readthedocs.io/en/latest/shared-bindings/audiocore/RawSample.html#audiocore.RawSample

Or, would it be more clear to change this param to `stereo`?

sommersoft commented on shared-bindings/audioio/__init__.c:
```diff
@@ -72,7 +75,7 @@ STATIC const mp_rom_map_elem_t audioio_module_globals_table[] = {
     { MP_ROM_QSTR(MP_QSTR_AudioOut), MP_ROM_PTR(&audioio_audioout_type) },
 #ifdef CIRCUITPY_AUDIOIO_COMPAT
-    { MP_ROM_QSTR(MP_QSTR_Mixer), MP_ROM_PTR(&audioio_mixer_type) },
+    //{ MP_ROM_QSTR(MP_QSTR_Mixer), MP_ROM_PTR(&audiomixer_mixer_type) },
```
As I went back to gather the compile error I was getting before, my non-late-at-night brain clicked. It was failing on boards that don't include `audiomixer`...which yeah, makes sense. 🤦‍♂

sommersoft commented:
This is ready for another look. I'm going to update the original PR comments with a summary for release notes.

ladyada commented:
@jepler please take a peek!

sommersoft commented on shared-module/audiomixer/Mixer.c:
```diff
@@ -194,32 +152,147 @@ uint32_t add16signed(uint32_t a, uint32_t b) {
+    #if (defined (__ARM_ARCH_7EM__) && (__ARM_ARCH_7EM__ == 1)) //Cortex-M4 w/FPU
+    return val;
+    #else*/
+    mp_printf(&mp_plat_print, "mult8unsigned");
```
Debug print snuck in! Will be removed in next push...

tannewt replied:
Please ping when it's gone. :-)

sommersoft replied:
Ping! 😄
